File: app/src/main/java/com/androidnerds/doordash/core/utils/Result.java
package com.androidnerds.doordash.core.utils;

public class Result<Data, Error> {
    
    private Data t;

    private Error e;
    
    public Result(Data t, Error e) {
        this.t = t;
        this.e = e;
    }

    public Data getData() {
        return t;
    }

    public Error getError() {
        return e;
    }
}

File: core/src/main/java/com/androidnerds/doordash/core/Result.kt
package com.androidnerds.doordash.core

sealed class Result<T, E> {

    data class Success<T>(val value: T) : Result<T, Nothing>()

    data class Failure<E>(val error: E) : Result<Nothing, E>()

    inline fun <C> fold(success: (T) -> C, failure: (E) -> C): C = when (this) {
        is Success -> success(value)
        is Failure -> failure(error)
    }

}

typealias SimpleResult<T> = Result<T, Throwable>